### DISCUSSION TITLE
any idea on how to stop request being proccessed to the next stage in middleware ?

### DISCUSSION THREAD

**Discussion by neomn** [fastapi/fastapi#10477] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.middleware("http")
async def add_session_id_middleware(request:Request, call_next):    
    if 'X-session-id' not in request.headers   and   'session_id' not in request.session:
        response = JSONResponse(content = {'set_session_id' : 'true'})
        response.headers['X-session-id'] = str(uuid4())
        return response
    response = await call_next(request)
    return response
```

### Description

i'm trying to return a custom JSON response if there is no session id in request headers and request session , but call_next is getting called regardless of the condition , why is that so ? 

### Operating System

Linux

### FastAPI Version

0.100.0

### Pydantic Version

2

### Python Version

3.11

**Comment by Kludex** [comment]
Can you provide a full example that I can run a single command to reproduce it?

**Comment by neomn** [comment]
@Kludex 
sure 

``` python
################ external imports  ############################
from uuid                          import uuid4
from fastapi                       import FastAPI, Request
from fastapi.responses             import JSONResponse
from starlette.middleware.sessions import SessionMiddleware


################# internal imports  ###########################
from .db.models.TempRegister import TempRegister
from .db.database            import SessionLocal, engin, Base
from .routes.v1              import  v1



Base.metadata.create_all(bind=engin)
app = FastAPI()
app.include_router(v1.app)


@app.middleware("http")
async def add_session_id_middleware(request:Request, call_next):    
    if 'X-session-id' not in request.headers   and   'session_id' not in request.session:
        response = JSONResponse(content = {'set_session_id' : 'true'})
        response.headers['X-session-id'] = str(uuid4())
        return response
    response = await call_next(request)
    return response


@app.get("/")
async def hello(request: Request):
    return ['home']


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()

        
app.add_middleware(SessionMiddleware, secret_key="some-random-string")
```

**Comment by jgould22** [comment]
I would start by logging or breaking/inspecting the values in your if to make sure everything is how you expect
```python
    if 'X-session-id' not in request.headers   and   'session_id' not in request.session:
```

**Comment by neomn** [comment, marked as answer]
@Kludex @jgould22 
well , i done some logging and eventhough i'm not sending any session_id  in request body , it turns out that apparently there is session_id in the session  , i'm using post man 

```Python
@app.middleware("http")
async def add_session_id_middleware(request:Request, call_next):    
    
    logging.debug(f"session_id_middleware> X-session-id is in request.headers > {'X-session-id' in request.headers}")
    logging.debug(f"session_id_middleware> session_id is in request.session> {'session_id' in request.session}")
   
    if 'X-session-id' not in request.headers   and   'session_id' not in request.session:
        logging.debug(f"session_id_middleware> adding X-session-id to response.headers")
        response = JSONResponse(content = {'set_session_id' : 'true'})
        response.headers['X-session-id'] = str(uuid4())
        return response

    logging.debug(f"session_id_middleware> call_next")
    response = await call_next(request)
    return response
```

> 2023-10-20 18:39:12 - root - DEBUG - session_id_middleware> X-session-id is in request.headers > False
2023-10-20 18:39:12 - root - DEBUG - session_id_middleware> session_id is in request.session> True
2023-10-20 18:39:12 - root - DEBUG - session_id_middleware> call_next
